### ISSUE TITLE
nio error  too many files open

### ISSUE THREAD

**Issue by chris-speebly** [grpc/grpc-swift#700]
### New Issue Checklist

- [ check] Updated grpc-swift to the latest version
- [ check] I read the [Contribution Guidelines](https://github.com/grpc/grpc-swift/blob/master/CONTRIBUTING.md)
- [ check] I searched for [existing GitHub issues](https://github.com/grpc/grpc-swift/issues)(especially, there are many [resolved issues related to SSL](https://github.com/grpc/grpc-swift/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+ssl))

### Issue Description

i receive this error inmy app

'try!' expression unexpectedly raised an error: kqueue() failed: Too many open files (errno: 24): file /Users/me/Library/Developer/Xcode/DerivedData/ClientSDK-gyptrzhdolqckggvdglkeegaxwka/SourcePackages/checkouts/swift-nio/Sources/NIO/EventLoop.swift, line 783

##### Complete output when running grpc-swift, including the stack trace and command used

<details>
  <pre>[INSERT OUTPUT HERE]</pre>
</details>

### Environment

<details>
  <pre>
  
| Key                           | Value                                                         |
| ----------------------------- | --------------------------------------------------------------|
| OS Version                    |   13.3                                                       |
| Swift Version                 |   5.1                                                       |
| Xcode Version                 |   11.3                                                       |
| gRPC-Swift Version            |   1.0.0-beta.8                                                       |
| gRPC-Swift Version            |   x.x.x                                                       |
| protoc Version                |   3                                                      |
| protoc-gen-swift Version      |   x.x.x                                                       |
| protoc-gen-swiftgrpc Version  |   x.x.x                                                       |
  </pre>
</details>

**Comment by glbrntt** [comment]
Can you provide more details on how you're using gRPC here? Are you shutting down your event loop group when you're done with it?

**Comment by chris-speebly** [comment]
sorry, i was able to resolve this error. This issue was caused by me.

**Comment by MrMage** [comment]
What did you have to change? Maybe it can help someone else in the future.

**Comment by chris-speebly** [comment]
the reason this error occurred was that I was calling the connection to the server in a loop. Once i removed the connection code outside of the loop the problem was resolved

**Comment by sabiland** [comment]
If it helps anyone: we have internal iOS app to test our automatic speech recognition (gRpc server) via gRpc. We found one case where our app was crashing. We were testing sequental tests.

After 256 tests, app crashes with: "too many files open". It turns out iPhone has only 256 network-file-descriptors available (which are released too slowly). We fixed that with: 

- Singleton MultiThreadedEventLoopGroup instance
- Singleton ClientConnection instance
- Singleton CallOptions instance

**Comment by weissi** [comment]
@sabiland Doing only a single `MultiThreadedEventLoopGroup` per application is _definitely the right choice_. Otherwise you're foregoing all the benefits of asynchronous programming. Only using a single `ClientConnection` also sounds like definitely the right choice because otherwise you're foregoing all the benefits of HTTP/2.

**Comment by MrMage** [comment]
It sound to me like the problem in that case was not multiple event loop groups running in parallel, but rather the corresponding network resources not getting released. Maybe there is a way to force-release the old network resources? Or maybe you had somehow still kept the old event loop groups around rather than shutting them down, effectively "leaking" them (or the corresponding file descriptors)?
